docs/architecture/adr-062-p2p-architecture.md

```diff
@@ -0,0 +1,531 @@
+# ADR 062: P2P Architecture and Abstractions
+
+## Changelog
+
+- 2020-11-09: Initial version (@erikgrinaker)
+
+## Context
+
+In [ADR 061](adr-061-p2p-refactor-scope.md) we decided to refactor the peer-to-peer (P2P) networking stack. The first phase is to redesign and refactor the internal P2P architecture, while retaining protocol compatibility as far as possible.
+
+## Alternative Approaches
+
+Several variations of the proposed design were considered, including e.g. calling interface methods instead of passing messages (like the current architecture), merging channels with streams, exposing the internal peer data structure to reactors, being message format-agnostic via arbitrary codecs, and so on. The final design was chosen because it has very loose coupling, is simpler to reason about and more convenient to use, avoids race conditions and lock contention for internal data structures, gives reactors better control of message ordering and processing semantics, and allows for QoS scheduling and backpressure in a very natural way.
```

Comment:
`The final design was chosen ...` What's the final design? Might be just me, but it reads a bit funny.

Reply:
Yeah, ADRs can be worded a bit weird since they're really meant to document a decision, so they need to be written as if the contents have been decided. But I suppose I can reword it a bit.